code/backend/batch/utilities/tools/QuestionAnswerTool.py

```diff
@@ -1,182 +1,113 @@
 import json
 import logging
+from typing import List
 import warnings
 from .AnsweringToolBase import AnsweringToolBase
 
-from langchain.chains.llm import LLMChain
-from langchain.prompts import (
-    AIMessagePromptTemplate,
-    ChatPromptTemplate,
-    FewShotChatMessagePromptTemplate,
-    HumanMessagePromptTemplate,
-    MessagesPlaceholder,
-    PromptTemplate,
-)
-from langchain_community.callbacks import get_openai_callback
-from langchain_core.documents import Document
-from langchain_core.messages import SystemMessage
-
 from ..helpers.AzureSearchHelper import AzureSearchHelper
 from ..helpers.ConfigHelper import ConfigHelper
 from ..helpers.LLMHelper import LLMHelper
 from ..helpers.EnvHelper import EnvHelper
 from ..common.Answer import Answer
 from ..common.SourceDocument import SourceDocument
+from azure.search.documents.models import VectorizedQuery
+
 
 logger = logging.getLogger(__name__)
 
 
 class QuestionAnswerTool(AnsweringToolBase):
+
     def __init__(self) -> None:
         self.name = "QuestionAnswer"
-        self.vector_store = AzureSearchHelper().get_vector_store()
-        self.verbose = True
+        self.azure_search_helper = AzureSearchHelper()
+        self.llm_helper = LLMHelper()
         self.env_helper = EnvHelper()
         self.config = ConfigHelper.get_active_config_or_default()
 
-    @staticmethod
-    def json_remove_whitespace(obj: str) -> str:
-        """
-        Remove whitespace from a JSON string.
-        """
-        try:
-            return json.dumps(json.loads(obj), separators=(",", ":"))
-        except json.JSONDecodeError:
-            return obj
-
-    def generate_llm_chain(self, question: str, sources: list[Document]):
-        answering_prompt = PromptTemplate(
-            template=self.config.prompts.answering_user_prompt,
-            input_variables=["question", "sources"],
-        )
+    def answer_question(
+        self, question: str, chat_history: List[dict], **kwargs: dict
+    ) -> Answer:
+        question_embeddings = self.llm_helper.generate_embeddings(question)
 
-        sources_text = "\n\n".join(
-            [f"[doc{i+1}]: {source.page_content}" for i, source in enumerate(sources)]
+        # Now that we are querying Azure Search using the official libraries, we can add and modify the query any way we need, such as adding a query for image embeddings
+        vector_query = VectorizedQuery(
+            vector=question_embeddings,
+            k_nearest_neighbors=self.env_helper.AZURE_SEARCH_TOP_K,
+            fields="content_vector",
+        )
+        search_results = self.azure_search_helper.search_client.search(
```

Comment:
This is the main reason for this change. It gives us full control of the query to AI Search, allowing us to supply multiple `VectorizedQuery`'s, one fo the text content and if enabled, one for the image content. We can then pass any image results to gpt-4 vision in the messages below